Lib/abc.py

```diff
@@ -121,6 +121,49 @@ def _abc_caches_clear(cls):
             """Clear the caches (for debugging or testing)."""
             _reset_caches(cls)
 
+def update_abstractmethods(cls):
+    """Repopulate the abstract methods of an abstract class, or a subclass of
+    an abstract class.
+
+    If a class has had one of its abstract methods implemented after the
+    class was created, the method will not be considered implemented until
+    this function is called. Alternatively, if a new abstract method has been
+    added to the class, it will only be considered an abstract method of the
+    class after this function is called.
+
+    This function should be called before any use is made of the class,
+    usually in class decorators that add methods to the subject class.
+
+    Returns cls, to allow usage as a class decorator.
+
+    If cls is has any subclasses, raises a RuntimeError.
+
+    If cls is not an instance of ABCMeta, does nothing.
+    """
+    if not hasattr(cls, '__abstractmethods__'):
+        # We check for __abstractmethods__ here because cls might by a C
+        # implementation or a python implementation (especially during
+        # testing), and we want to handle both cases.
+        return cls
+
+    if cls.__subclasses__():
+        raise RuntimeError("cannot update abstract methods of class after"
+                           " subclassing")
+
+    abstracts = set()
+    # Check the existing abstract methods of the parents, keep only the ones
+    # that are not implemented.
+    for scls in cls.__bases__:
+        for name in getattr(scls, '__abstractmethods__', ()):
+            value = getattr(cls, name, None)
+            if getattr(value, "__isabstractmethod__", False):
+                abstracts.add(name)
```

Comment:
Add tests for deeper inheritance - I think iterating over `__bases__` may not be enough:

```
>>> class A: pass
... 
>>> class B(A): pass
... 
>>> class C(B): pass
... 
>>> C.__bases__
(<class '__main__.B'>,)
>>> B.__bases__
(<class '__main__.A'>,)
```

Reply:
It actually is- we are assured that the parent's `__abstractmethods__` are correct since, after a class is subclassed, its abstraction status cannot change:
```
class A(ABC):
  @abstractmethod
  def foo(self): pass

class B(A):
  pass

# after this point, A.__abstractmethods__ will never change, so we can rely on it always being consistent
```

Reply:
I added a test case regardless

Reply:
I see, interesting. How about this case? I think you might need to break after the first hit for the method:

```
class A(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def foo(self):
        pass

class B(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def foo(self):
        pass

class C(A, B):
    pass

A.foo = lambda self: None


assert(C.foo is not abstract)
```

Reply:
This is expected (albeit confusing) behavior. Since abstraction status cannot change after subclassing, A is still considered abstract, even though its method is implemented. However, since namespace resolution occurs through the `__dict__` attribute, The method is defined as far as C is concerned. Indeed, if you were to call `update_abstractmethods(C)` after this code, C would become concrete and usable, since its `foo` method implemented.

In essence, abstract methods shouldn't be changed after the class is used, so this use case is not handled by this implementation.

Reply:
I see. I was concerned about something a bit different, but the test I suggested didn't quite capture that.

See the code below. I wondered if the last update_abstractmethods will know to ignore A and look only at B. It works because you put 'value = getattr(cls, name, None)'  (cls and not scls). So I guess it checks this once for A and once for B, but gets the same (correct) result both times.

Might be worth adding this test to make sure nobody breaks that getattr. 

```
import abc

class A(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def foo(self):
        pass

class B(metaclass=abc.ABCMeta):
    def foo(self):
        pass

class C(B, A):
    @abc.abstractmethod
    def foo(self):
        pass

abc.update_abstractmethods(C)

del C.foo

abc.update_abstractmethods(C)
```

Reply:
Well, I disagree with the check for empty `__subclasses__()`, and I think in Irit's last example the right thing to do is to call `abc.update_abstractmethods(A)` and then `abc.update_abstractmethods(C)`.

Note that it's not technically true that the abstractness cannot change once subclassed -- it may be undocumented but one could just write `A.__abstractmethods__ = ()`.

FWIW I do agree that the code should *assume* that the abstractness of the bases is correct. If some base lies about it, well, Garbage In, Garbage Out.

Reply:
I don't think abc.update_abstractmethods(A) would have any impact here, A never changed.

Reply:
Sorry, I was commenting on your earlier example which has `A.foo = lambda ...` -- in that example it certainly would have an effect. :-) And my disagreement was with this quote from @bentheiii:

> In essence, abstract methods shouldn't be changed after the class is used, so this use case is not handled by this implementation.

Reply:
> Sorry, I was commenting on your earlier example which has `A.foo = lambda ...` -- in that example it certainly would have an effect. :-)
> 

Ah, in that case it raises an exception that A has been subclasses so you can't update it anymore. I think there's a test for that already but if not then there should be.

Reply:
I see now - you're suggesting to remove that exception. In that case this test is probably needed.